File: src/core/gui/Select.ts
import { LabelledControl, LabelledControlOptions, LabelPosition } from "./LabelledControl";
import { disposable_listener, div, Icon } from "./dom";
import "./Select.css";
import { is_property, Property } from "../observable/property/Property";
import { Button } from "./Button";
import { WritableProperty } from "../observable/property/WritableProperty";
import { WidgetProperty } from "../observable/property/WidgetProperty";
import { Menu } from "./Menu";
import { property } from "../observable";

export type SelectOptions<T> = LabelledControlOptions & {
    readonly items: readonly T[] | Property<readonly T[]>;
    readonly to_label?: (element: T) => string;
    readonly selected?: T | Property<T>;
};

export class Select<T> extends LabelledControl {
    readonly element = div({ className: "core_Select" });

    readonly preferred_label_position: LabelPosition;

    readonly selected: WritableProperty<T | undefined>;

    private readonly items: Property<readonly T[]>;
    private readonly to_label: (element: T) => string;
    private readonly button: Button;
    private readonly menu: Menu<T>;
    private readonly _selected: WidgetProperty<T | undefined>;
    private just_opened: boolean;

    constructor(options: SelectOptions<T>) {
        super(options);

        this.preferred_label_position = "left";

        this.items = is_property(options.items) ? options.items : property(options.items);
        this.to_label = options.to_label ?? String;
        this.button = this.disposable(
            new Button({
                text: " ",
                icon_right: Icon.TriangleDown,
            }),
        );
        this.menu = this.disposable(
            new Menu<T>({
                items: this.items,
                to_label: this.to_label,
                related_element: this.element,
            }),
        );
        this.element.append(this.button.element, this.menu.element);

        this._selected = new WidgetProperty<T | undefined>(this, undefined, this.set_selected);
        this.selected = this._selected;

        this.just_opened = false;

        this.disposables(
            disposable_listener(this.button.element, "mousedown", this.button_mousedown),

            this.button.onmouseup.observe(this.button_mouseup),

            this.button.onkeydown.observe(this.button_keydown),

            this.menu.selected.observe(({ value }) => {
                this._selected.set_val(value, { silent: false });
            }),
        );

        if (options) {
            if (is_property(options.selected)) {
                this.selected.bind_to(options.selected);
            } else if (options.selected) {
                this.selected.val = options.selected;
            }
        }

        this.finalize_construction();
    }

    protected set_enabled(enabled: boolean): void {
        super.set_enabled(enabled);
        this.button.enabled.val = enabled;
    }

    protected set_selected(selected?: T): void {
        this.button.text.val = selected !== undefined ? this.to_label(selected) : " ";
        this.menu.selected.val = selected;
    }

    private button_mousedown = (e: Event): void => {
        e.stopPropagation();
        this.just_opened = !this.menu.visible.val;
        this.menu.visible.val = true;
    };

    private button_mouseup = (): void => {
        if (this.just_opened) {
            this.menu.focus();
        } else {
            this.menu.visible.val = false;
        }

        this.just_opened = false;
    };

    private button_keydown = ({ value: evt }: { value: KeyboardEvent }): void => {
        switch (evt.key) {
            case "Enter":
            case " ":
                evt.preventDefault();
                this.just_opened = !this.menu.visible.val;
                this.menu.visible.val = true;
                this.menu.focus();
                this.menu.hover_next();
                break;

            case "ArrowUp":
                {
                    if (this._selected.val === undefined) break;

                    const index = this.items.val.indexOf(this._selected.val) - 1;
                    if (index < 0) break;

                    this._selected.set_val(this.items.val[index], { silent: false });
                }
                break;

            case "ArrowDown":
                {
                    if (this._selected.val === undefined) break;

                    const index = this.items.val.indexOf(this._selected.val) + 1;
                    if (index >= this.items.val.length) break;

                    this._selected.set_val(this.items.val[index], { silent: false });
                }
                break;
        }
    };
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Select.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.Node
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.events.MouseEvent
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.list.emptyListCell
import world.phantasmal.observable.cell.mutableCell
import world.phantasmal.observable.cell.nullCell
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.dom.Icon
import world.phantasmal.webui.dom.div

class Select<T : Any>(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    tooltip: Cell<String?> = nullCell(),
    private val className: String? = null,
    label: String? = null,
    labelCell: Cell<String>? = null,
    preferredLabelPosition: LabelPosition = LabelPosition.Before,
    private val items: Cell<List<T>> = emptyListCell(),
    private val itemToString: (T) -> String = Any::toString,
    private val selected: Cell<T?> = nullCell(),
    private val onSelect: (T) -> Unit = {},
) : LabelledControl(visible, enabled, tooltip, label, labelCell, preferredLabelPosition) {
    private val buttonText = mutableCell(" ")

    private lateinit var menu: Menu<T>
    private val menuVisible = mutableCell(false)
    private var justOpened = false

    override fun Node.createElement() =
        div {
            className = "pw-select"

            this@Select.className?.let { classList.add(it) }

            // Default to a single space so the inner text part won't be hidden.
            observeNow(selected) { buttonText.value = it?.let(itemToString) ?: " " }

            addWidget(Button(
                enabled = enabled,
                textCell = buttonText,
                iconRight = Icon.TriangleDown,
                onMouseDown = ::onButtonMouseDown,
                onMouseUp = { onButtonMouseUp() },
                onKeyDown = ::onButtonKeyDown,
            )).element.id = labelId
            menu = addWidget(Menu(
                visible = menuVisible,
                enabled = enabled,
                items = items,
                itemToString = itemToString,
                onSelect = ::select,
                onCancel = { menuVisible.value = false },
            ))
        }

    private fun onButtonMouseDown(e: MouseEvent) {
        e.stopPropagation()
        justOpened = !menuVisible.value
        menuVisible.value = true
        selected.value?.let(menu::highlightItem)
    }

    private fun onButtonMouseUp() {
        if (justOpened) {
            menu.focus()
        } else {
            menuVisible.value = false
        }

        justOpened = false
    }

    private fun onButtonKeyDown(e: KeyboardEvent) {
        when (e.key) {
            "ArrowDown" -> {
                if (items.value.isNotEmpty()) {
                    if (selected.value == null) {
                        select(items.value.first())
                    } else {
                        val index = items.value.indexOf(selected.value) + 1

                        if (index >= items.value.size) {
                            select(items.value.first())
                        } else {
                            select(items.value[index])
                        }
                    }
                }
            }

            "ArrowUp" -> {
                if (items.value.isNotEmpty()) {
                    if (selected.value == null) {
                        select(items.value.last())
                    } else {
                        val index = items.value.indexOf(selected.value) - 1

                        if (index < 0) {
                            select(items.value.last())
                        } else {
                            select(items.value[index])
                        }
                    }
                }
            }

            "Enter", " " -> {
                e.preventDefault()
                e.stopPropagation()

                justOpened = !menuVisible.value
                menuVisible.value = true
                selected.value?.let(menu::highlightItem)
                menu.focus()
            }
        }
    }

    private fun select(item: T) {
        menuVisible.value = false
        onSelect(item)
    }

    companion object {
        init {
            @Suppress("CssUnusedSymbol")
            // language=css
            style("""
                .pw-select {
                    position: relative;
                    display: inline-flex;
                    width: 160px;
                }

                .pw-select .pw-button {
                    flex: 1;
                }

                .pw-select .pw-menu {
                    top: 23px;
                    left: 0;
                    min-width: 100%;
                }
            """.trimIndent())
        }
    }
}
